tests/baselines/reference/api/typescript.d.ts

```diff
@@ -3854,6 +3838,258 @@ declare namespace ts {
+        /**
+         * Reports the diagnostics in reading/writing or parsing of the config file
+         */
+        onConfigFileDiagnostic: DiagnosticReporter;
+        /**
+         * Reports unrecoverable error when parsing config file
+         */
+        onUnRecoverableConfigFileDiagnostic: DiagnosticReporter;
+    }
+    /**
+     * Host to create watch with config file
+     */
+    interface WatchCompilerHostOfConfigFile extends WatchCompilerHost, ConfigFileDiagnosticsReporter {
+        /** Name of the config file to compile */
+        configFileName: string;
+        /** Options to extend */
+        optionsToExtend?: CompilerOptions;
+        /**
+         * Used to generate source file names from the config file and its include, exclude, files rules
+         * and also to cache the directory stucture
+         */
+        readDirectory(path: string, extensions?: ReadonlyArray<string>, exclude?: ReadonlyArray<string>, include?: ReadonlyArray<string>, depth?: number): string[];
+    }
+    interface Watch<T> {
+        /** Synchronize with host and get updated program */
+        getProgram(): T;
+    }
+    /**
+     * Creates the watch what generates program using the config file
+     */
+    interface WatchOfConfigFile<T> extends Watch<T> {
+    }
+    /**
+     * Creates the watch that generates program using the root files and compiler options
+     */
+    interface WatchOfFilesAndCompilerOptions<T> extends Watch<T> {
+        /** Updates the root files in the program, only if this is not config file compilation */
+        updateRootFileNames(fileNames: string[]): void;
+    }
+    /**
+     * Create the watched program for config file
+     */
+    function createWatchOfConfigFile(configFileName: string, optionsToExtend?: CompilerOptions, system?: System, reportDiagnostic?: DiagnosticReporter, reportWatchStatus?: WatchStatusReporter): WatchOfConfigFile<Program>;
+    /**
+     * Create the watched program for root files and compiler options
+     */
+    function createWatchOfFilesAndCompilerOptions(rootFiles: string[], options: CompilerOptions, system?: System, reportDiagnostic?: DiagnosticReporter, reportWatchStatus?: WatchStatusReporter): WatchOfFilesAndCompilerOptions<Program>;
+    /**
+     * Creates the watch from the host for root files and compiler options
+     */
+    function createWatchProgram(host: WatchCompilerHostOfFilesAndCompilerOptions): WatchOfFilesAndCompilerOptions<Program>;
+    /**
+     * Creates the watch from the host for config file
+     */
+    function createWatchProgram(host: WatchCompilerHostOfConfigFile): WatchOfConfigFile<Program>;
+    /**
+     * Creates the watch from the host for root files and compiler options
+     */
+    function createWatchBuilderProgram<T extends BuilderProgram>(host: WatchCompilerHostOfFilesAndCompilerOptions & BuilderProgramHost, createBuilderProgram: (newProgram: Program, host: BuilderProgramHost, oldProgram?: T) => T): WatchOfFilesAndCompilerOptions<T>;
```

Comment:
seems that this is the default behavior of `createWatchProgram` anyways, and to override that you need to change the `onAfterProgramCreate` on the host.. if this is accurate, why do we need this helper?

Reply:
```createWatchProgram```  just creates the program.. The user of ```createWatchProgram``` gets the Program. ```createWatchBuilderProgram``` creates the watchProgram and exposes the builder program instead of watchProgram. Eg. [ts-loader](https://github.com/TypeStrong/ts-loader/pull/685/files#diff-f48cc90ddff8a9e808d82e53f353649bR170) doesnt need builder at all.. But [gulp-tsb](https://github.com/jrieken/gulp-tsb/pull/74/files#diff-f8924a760912596517f9ae8931e4f7a5R166) uses createBuilderProgram

Reply:
in `createWatchCompilerHost` we return an object whose `afterProgramCreate` property is set to `emitFilesAndReportErrorUsingBuilder` by default. this will create a builderProgram using `createEmitAndSemanticDiagnosticsBuilderProgram`. what am i missing?

Reply:
But ts-loader or gulp-tsb do not use createWatchCompilerHost to create watch compiler host. Instead they create their own WatchCompilerHost that doesnt have these properties set.